### DISCUSSION TITLE
Node classification on HeteroData() (Heterogeneous graph)

### DISCUSSION THREAD

**Discussion by maryamhgf** [pyg-team/pytorch_geometric#7092] [Q&A] [answered]
Hi,

I have a dataset which is representing the computational graph of DNN models. I am using HeteroData().
Data:
Nodes: operators (Conv, Add, etc.)
Edges: the relationship between two operators ('Constant', 'connects', 'Add')
An overview of the data structure:
`HeteroData(
  Constant={ x=[301, 8] },
  Add={ x=[211, 5] },
  Mul={ x=[203, 5] },
  Conv={ x=[156, 20] },
  Transpose={ x=[125, 10] },
  ReduceMean={ x=[125, 7] },
  Div={ x=[114, 5] },
  Sub={ x=[64, 5] },
  Pow={ x=[65, 5] },
  Sqrt={ x=[66, 5] },
  Erf={ x=[53, 5] },
  GlobalAveragePool={ x=[3, 5] },
  Flatten={ x=[3, 6] },
  Gemm={ x=[3, 8] },
  (Constant, connects, Constant)={ edge_index=[2, 283] },
  (Constant, connects, Add)={ edge_index=[2, 23] },
  (Constant, connects, Conv)={ edge_index=[2, 1] },
  (Conv, connects, Mul)={ edge_index=[2, 21] },
  (Mul, connects, Add)={ edge_index=[2, 23] },
  (Add, connects, Conv)={ edge_index=[2, 9] },
  (Add, connects, Add)={ edge_index=[2, 9] },
  (Conv, connects, Transpose)={ edge_index=[2, 10] },
  (Transpose, connects, ReduceMean)={ edge_index=[2, 12] },
  (Transpose, connects, Sub)={ edge_index=[2, 12] },
...
  (Conv, connects, Add)={ edge_index=[2, 1] },
  (Transpose, connects, GlobalAveragePool)={ edge_index=[2, 1] },
  (GlobalAveragePool, connects, Flatten)={ edge_index=[2, 1] },
  (Flatten, connects, Gemm)={ edge_index=[2, 1] }
)`

As can be seen, the features of each node type are different from the other ones. I want to add a node and based on its connections, predict its type (node classification). Here type is the operator type (Add, Conv, etc.).  

I am not sure which network structure I should use. I appreciate any suggestions, thoughts, etc. I have tried HeteroGNN. But, since the features are different for each node type, I am not sure how I can implement it correctly.

**Comment by wsad1** [comment, marked as answer]
What do you mean by `But, since the features are different for each node type`. Do you mean the feature dimension is different for each node type? If so heterogenous GNNs should be able to handle node types with different feature dimension.
Check out this [example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_conv_dblp.py).
